### PR TITLE
Add query item limit

### PR DESCRIPTION
**Pull request by olaaun**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: container-search/src/main/java/com/yahoo/prelude/query/QueryCanonicalizer.java [hunk 1]
```diff
@@ -19,25 +20,38 @@
...
      * Canonicalizes this query
      * 
      * @return null if the query is valid, an error message if it is invalid
      */
-    public static String canonicalize(QueryTree query) {
+    private static String canonicalize(QueryTree query, Integer maxQueryItems) {
         ListIterator<Item> rootItemIterator = query.getItemIterator();
         CanonicalizationResult result = recursivelyCanonicalize(rootItemIterator.next(), rootItemIterator);
         if (query.isEmpty() && ! result.isError()) result = CanonicalizationResult.error("No query");
+        if (query.getTreeSize() > maxQueryItems) result = CanonicalizationResult.error("Query tree exceeds allowed item count");
```
Comment (geirst) [inline container-search/src/main/java/com/yahoo/prelude/query/QueryCanonicalizer.java:54]: Please add configured item count to the error message.

**Comment by baldersheim** [comment]
@geirst PTAL

File: container-search/src/main/java/com/yahoo/search/query/QueryTree.java [hunk 2]
```diff
@@ -179,4 +180,23 @@ private static void getPositiveTerms(Item item, List<IndexedItem> terms) {
         }
     }
 
+    /**
+     * @return The total number of items in this query tree
```
Comment (bratseth) [inline container-search/src/main/java/com/yahoo/search/query/QueryTree.java:184]:
```suggestion
     * Returns the total number of items in this query tree.
```

File: container-search/src/main/java/com/yahoo/search/query/QueryTree.java [hunk 3]
```diff
@@ -179,4 +180,23 @@ private static void getPositiveTerms(Item item, List<IndexedItem> terms) {
...
+     */
+    public int getTreeSize() {
```
Comment (bratseth) [inline container-search/src/main/java/com/yahoo/search/query/QueryTree.java:186]: This isn't a getter so I think you should use treeSize(), or maybe countItems() to communicate that this isn't cheap.

**Review by bratseth — APPROVED** [review]
👍

**Pull request merged**